GPS traces support

BlueBird931 opened the pull request.

BlueBird931 committed:
Reintegrated gps_trace files

BlueBird931 committed:
added coordinates to meters conversion (to fix)

BlueBird931 committed:
Fixed coordinate conversion algorithm

BlueBird931 committed:
Added position offset

Also started implementing timestamp loading from gpx file.
Current function requires fixing.

BlueBird931 committed:
Finalized timestamp management

BlueBird931 committed:
Started defining "follow_track" method in gps_trace.hpp

BlueBird931 committed:
Implemented follow_track method in gps_trace.hpp

BlueBird931 committed:
Added owner_node_uid check to gps_trace

BlueBird931 committed:
Fixed conversion bug in gps_trace::next_point()

Also removed old test file

BlueBird931 committed:
Replaced simple offset with ref_lat/lon in gps_trace

BlueBird931 committed:
Minor fixes and comments

BlueBird931 committed:
Started final modifications

- Moved the gps_trace classes from src\lib\gps to src\lib\simulation
- Started reworking next_point method, to be finished
- Updated class data structures

BlueBird931 committed:
Finished next_node() modifications in gps_trace

BlueBird931 committed:
Modifiche a seguito di colloquio

BlueBird931 committed:
partial fix for gps_trace crash

BlueBird931 committed:
gps_trace refactoring and debugging

BlueBird931 committed:
Fixed times_t casting and added comments

Harniver commented on .gitignore:
```diff
@@ -9,3 +9,4 @@ output/
 bazel-*
 _bazel_*
 node_net.md
+/src/lib/gps/build
```
Questa modifica direi che non ha senso in generale, probabilmente era servita a te in locale ma la toglierei

Harniver commented on src/CMakeLists.txt:
```diff
@@ -198,6 +198,7 @@ set(
     lib/simulation/simulated_map.cpp
     lib/simulation/simulated_positioner.cpp
     lib/simulation/spawner.cpp
+    lib/simulation/gps_trace.cpp
```
Questi file vengono tenuti in ordine alfabetico, ma prima di riordinare qui, io sposterei i file di gps_trace dalla cartella simulation/ (che contiene componenti FCPP per la simulazione) alla cartella data/ (che contiene tipi di dato che vengono usati dai programmi aggregati)

Harniver commented on src/lib/simulation.hpp:
```diff
@@ -17,6 +17,7 @@
 #include "lib/simulation/simulated_positioner.hpp"
 #include "lib/simulation/simulated_map.hpp"
 #include "lib/simulation/spawner.hpp"
+#include "lib/simulation/gps_trace.hpp"
```
Anche in questi file si tiene l'ordine alfabetico (ma come scritto prima sposterei la gps_trace, quindi la modifica in ogni caso finirà sul file data.hpp)

Harniver commented on src/lib/simulation/gps_trace.cpp:
```diff
@@ -0,0 +1,123 @@
+#include "gps_trace.hpp"
+
+#define EARTH_RADIUS 6371000.0
```
Invece di define, `static constexpr double EARTH_RADIUS = ...;`

Harniver commented on src/lib/simulation/gps_trace.cpp:
```diff
@@ -0,0 +1,123 @@
+#define EARTH_RADIUS 6371000.0
+
+namespace fcpp
+{
```
Nella convenzione usata in FCPP, le graffe aperte non vanno a capo ma sulla stessa linea (qui e ovunque sotto accada)

Harniver commented on src/lib/simulation/gps_trace.cpp:
```diff
@@ -0,0 +1,123 @@
+{
+
+gps_trace::gps_trace(const std::string &src_gpx_file, const double ref_lat, const double ref_lon, const times_t ref_time) {
+    std::cout << "gps_trace constructor - 1" << std::endl;
```
Stampa di debug da rimuovere

Harniver commented on src/lib/simulation/gps_trace.cpp:
```diff
@@ -0,0 +1,123 @@
+        }
+
+        for (rapidxml::xml_node<> *trk_node = gpx_node->first_node("trk");
+                trk_node;
```
Qui vedo errori di indentazione, probabilmente per via dei soliti problemi tab/spazi. Lo standard in FCPP è di indentare con 4 spazi, prova a impostare il tuo editor in questo modo e ri-salvare il file e dovrebbe sistemare da solo

Harniver commented on src/lib/simulation/gps_trace.hpp:
```diff
@@ -0,0 +1,144 @@
+     * @param ref_lat reference latitude mapped at in x:0
+     * @param ref_lon reference longitude mapped at in y:0
+     */
+    vec<2> coord_to_meters(double lat, double lon, double ref_lat, double ref_lon); // private static
```
Questo e quello sotto dovrebbero essere private e static (l'hai anche messo come commento)

Harniver commented on src/lib/simulation/gps_trace.hpp:
```diff
@@ -0,0 +1,144 @@
+
+    /**
+     *! @brief aggregate function to follow the GPS track assigned to the calling node
+     */
```
Questa funzione la vedo più come una funzione aggregata che segue una traccia GPS, quindi la sposterei da qui al file lib/coordination/geometry.hpp, aggiungendo come parametro un riferimento a oggetto gps_trace. Questo probabilmente richiederà di aggiungere qualche funzione di interfaccia all'oggetto gps_trace (per esempio per fare find() di una delle tracce)

Harniver reviewed (changes requested):
Good PR, but there are a few refactoring and minor fixes to be made.
Also, some tests should be added, and then we can go for a final review.

BlueBird931 committed:
Refactoring for PR review

Harniver commented on src/lib/coordination/geometry.hpp:
```diff
@@ -239,6 +240,40 @@ inline typename node_t::position_type neighbour_charged_force(node_t& node, trac
+
+    double velocity;
+    if (time_left > next_time_step)
+    {
```
graffa a capo

Harniver commented on src/lib/data/gps_trace.hpp:
```diff
@@ -0,0 +1,105 @@
+    }
+
+private:
+    std::unordered_map<device_t, track_data> tracks;
```
rinominare tracks in m_tracks

Harniver commented on src/lib/data/gps_trace.hpp:
```diff
@@ -0,0 +1,105 @@
+     * @param ref_time Time offset for track timestamps.
+     * @param uid Uid of the node that will follow the track.
+     */
+    gps_trace(const std::string &src_gpx_file, const double ref_lat, const double ref_lon, const times_t ref_time);
```
fare due versioni del costruttore, una con questa signature che prende come argomento direttamente la stringa XML, e una con argomento `const char *` che prende l'indirizzo din un file, ne legge il content e richiama l'altro costruttore

Harniver commented on .gitignore:
```diff
@@ -8,4 +8,4 @@ output/
 bazel-*
 _bazel_*
-node_net.md
+node_net.md
```
forse c'è una riga vuota?

Harniver commented on src/lib/coordination/geometry.hpp:
```diff
@@ -12,6 +12,7 @@
 
 #include "lib/coordination/utils.hpp"
 #include "lib/data/vec.hpp"
+#include "lib/data/gps_trace.hpp"
```
invertire per ordine alfabetico

Harniver commented on src/lib/coordination/geometry.hpp:
```diff
@@ -239,6 +240,40 @@ inline typename node_t::position_type neighbour_charged_force(node_t& node, trac
+    node.velocity() = direction * velocity;
+
+    return true;
+}
```
Dopo la funzione aggiungi anche:
```
//! @brief Export list for follow_track.
using follow_track_t = common::export_list<>;
```
Ho anche una domanda: se uno riutilizza la gps_trace in una seconda simulazione, il sistema lo capisce e resetta l'index all'inizio? Avrebbe senso mettere anche un test che verifica che la gps_trace funziona quando la si riusa una seconda volta.

Harniver commented on src/lib/coordination/geometry.hpp:
```diff
@@ -239,6 +240,40 @@ inline typename node_t::position_type neighbour_charged_force(node_t& node, trac
+bool follow_track(node_t &node, trace_t call_point, gps_trace& trace) {
+    track_data td = trace.find_track(node.uid);
+
+    if (td.index == -1) { return false; } //track not found for current node
```
in caso di one-liner, non servono le graffe

Harniver commented on src/lib/data/gps_trace.cpp:
```diff
@@ -0,0 +1,123 @@
+    return tp;
+}
+
+track_data gps_trace::find_track(device_t uid) {
```
attenzione che questo fa una copia del track_data, che mi sa che non è voluto... probabilmente vuoi restituire un `track_data&` (anche se questo vuol dire tenersi da qualche parte nella classe un campo `static track_data s_empty_track{{}, -1};` da restituire quando non si trova la traccia)

Harniver reviewed (changes requested):
Getting closer but still some things to check remain!

BlueBird931 committed:
Preparation for tests

Added new constructor with XML string parameter instead of src.

Encapsulated parsing logic in new init() method to be used by both constructors.

General refactoring to address coding standards.

BlueBird931 committed:
Started adding tests for gps_trace

Added index resetting when reaching the end of a track.
Changed next_point and find_track methods to return object references instead of copies.
Improved comments
Added first test for gps_trace

BlueBird931 committed:
Added gps_trace tests

Improved parsing test
Added Navigation test to check the next_point method behavior.
Added tests for invalid GPX files and valid but empty GPX files.

BlueBird931 committed:
Add elevation support to gps_trace parsing

Added elevation handling in gps_trace by updating constructors, parsing logic, and updated track_point structure.
Updated test files to verify correct elevation parsing and conversion.

Harniver committed:
Merge branch 'dev' into gps_trace

Harniver committed:
general polishing & fixing docs

Harniver committed:
refactoring follow_track and next_point to avoid indexes in track data

Harniver committed:
refactoring gps_trace constructors to add a reference uid and time

Harniver committed:
nitpicking
